File: frontend/ordem_servico/lib/app/data/model/os_abertura_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:ordem_servico/app/infra/infra_imports.dart';
import 'package:ordem_servico/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class OsAberturaModel {
	int? id;
	int? idOsStatus;
	int? idColaborador;
	int? idCliente;
	String? numero;
	DateTime? dataInicio;
	String? horaInicio;
	DateTime? dataPrevisao;
	String? horaPrevisao;
	DateTime? dataFim;
	String? horaFim;
	String? nomeContato;
	String? foneContato;
	String? observacaoCliente;
	String? observacaoAbertura;
	List<OsAberturaEquipamentoModel>? osAberturaEquipamentoModelList;
	List<OsProdutoServicoModel>? osProdutoServicoModelList;
	List<OsEvolucaoModel>? osEvolucaoModelList;
	ViewPessoaClienteModel? viewPessoaClienteModel;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	OsStatusModel? osStatusModel;

	OsAberturaModel({
		this.id,
		this.idOsStatus,
		this.idColaborador,
		this.idCliente,
		this.numero,
		this.dataInicio,
		this.horaInicio,
		this.dataPrevisao,
		this.horaPrevisao,
		this.dataFim,
		this.horaFim,
		this.nomeContato,
		this.foneContato,
		this.observacaoCliente,
		this.observacaoAbertura,
		this.osAberturaEquipamentoModelList,
		this.osProdutoServicoModelList,
		this.osEvolucaoModelList,
		this.viewPessoaClienteModel,
		this.viewPessoaColaboradorModel,
		this.osStatusModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero',
		'data_inicio',
		'hora_inicio',
		'data_previsao',
		'hora_previsao',
		'data_fim',
		'hora_fim',
		'nome_contato',
		'fone_contato',
		'observacao_cliente',
		'observacao_abertura',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero',
		'Data Inicio',
		'Hora Inicio',
		'Data Previsao',
		'Hora Previsao',
		'Data Fim',
		'Hora Fim',
		'Nome Contato',
		'Fone Contato',
		'Observacao Cliente',
		'Observacao Abertura',
	];

	OsAberturaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idOsStatus = jsonData['idOsStatus'];
		idColaborador = jsonData['idColaborador'];
		idCliente = jsonData['idCliente'];
		numero = jsonData['numero'];
		dataInicio = jsonData['dataInicio'] != null ? DateTime.tryParse(jsonData['dataInicio']) : null;
		horaInicio = jsonData['horaInicio'];
		dataPrevisao = jsonData['dataPrevisao'] != null ? DateTime.tryParse(jsonData['dataPrevisao']) : null;
		horaPrevisao = jsonData['horaPrevisao'];
		dataFim = jsonData['dataFim'] != null ? DateTime.tryParse(jsonData['dataFim']) : null;
		horaFim = jsonData['horaFim'];
		nomeContato = jsonData['nomeContato'];
		foneContato = jsonData['foneContato'];
		observacaoCliente = jsonData['observacaoCliente'];
		observacaoAbertura = jsonData['observacaoAbertura'];
		osAberturaEquipamentoModelList = (jsonData['osAberturaEquipamentoModelList'] as Iterable?)?.map((m) => OsAberturaEquipamentoModel.fromJson(m)).toList() ?? [];
		osProdutoServicoModelList = (jsonData['osProdutoServicoModelList'] as Iterable?)?.map((m) => OsProdutoServicoModel.fromJson(m)).toList() ?? [];
		osEvolucaoModelList = (jsonData['osEvolucaoModelList'] as Iterable?)?.map((m) => OsEvolucaoModel.fromJson(m)).toList() ?? [];
		viewPessoaClienteModel = jsonData['viewPessoaClienteModel'] == null ? ViewPessoaClienteModel() : ViewPessoaClienteModel.fromJson(jsonData['viewPessoaClienteModel']);
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		osStatusModel = jsonData['osStatusModel'] == null ? OsStatusModel() : OsStatusModel.fromJson(jsonData['osStatusModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idOsStatus'] = idOsStatus != 0 ? idOsStatus : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['idCliente'] = idCliente != 0 ? idCliente : null;
		jsonData['numero'] = numero;
		jsonData['dataInicio'] = dataInicio != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInicio!) : null;
		jsonData['horaInicio'] = horaInicio;
		jsonData['dataPrevisao'] = dataPrevisao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataPrevisao!) : null;
		jsonData['horaPrevisao'] = Util.removeMask(horaPrevisao);
		jsonData['dataFim'] = dataFim != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFim!) : null;
		jsonData['horaFim'] = Util.removeMask(horaFim);
		jsonData['nomeContato'] = nomeContato;
		jsonData['foneContato'] = Util.removeMask(foneContato);
		jsonData['observacaoCliente'] = observacaoCliente;
		jsonData['observacaoAbertura'] = observacaoAbertura;
		
		var osAberturaEquipamentoModelLocalList = []; 
		for (OsAberturaEquipamentoModel object in osAberturaEquipamentoModelList ?? []) { 
			osAberturaEquipamentoModelLocalList.add(object.toJson); 
		}
		jsonData['osAberturaEquipamentoModelList'] = osAberturaEquipamentoModelLocalList;
		
		var osProdutoServicoModelLocalList = []; 
		for (OsProdutoServicoModel object in osProdutoServicoModelList ?? []) { 
			osProdutoServicoModelLocalList.add(object.toJson); 
		}
		jsonData['osProdutoServicoModelList'] = osProdutoServicoModelLocalList;
		
		var osEvolucaoModelLocalList = []; 
		for (OsEvolucaoModel object in osEvolucaoModelList ?? []) { 
			osEvolucaoModelLocalList.add(object.toJson); 
		}
		jsonData['osEvolucaoModelList'] = osEvolucaoModelLocalList;
		jsonData['viewPessoaClienteModel'] = viewPessoaClienteModel?.toJson;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['osStatusModel'] = osStatusModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idOsStatus = plutoRow.cells['idOsStatus']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		idCliente = plutoRow.cells['idCliente']?.value;
		numero = plutoRow.cells['numero']?.value;
		dataInicio = Util.stringToDate(plutoRow.cells['dataInicio']?.value);
		horaInicio = plutoRow.cells['horaInicio']?.value;
		dataPrevisao = Util.stringToDate(plutoRow.cells['dataPrevisao']?.value);
		horaPrevisao = plutoRow.cells['horaPrevisao']?.value;
		dataFim = Util.stringToDate(plutoRow.cells['dataFim']?.value);
		horaFim = plutoRow.cells['horaFim']?.value;
		nomeContato = plutoRow.cells['nomeContato']?.value;
		foneContato = plutoRow.cells['foneContato']?.value;
		observacaoCliente = plutoRow.cells['observacaoCliente']?.value;
		observacaoAbertura = plutoRow.cells['observacaoAbertura']?.value;
		osAberturaEquipamentoModelList = [];
		osProdutoServicoModelList = [];
		osEvolucaoModelList = [];
		viewPessoaClienteModel = ViewPessoaClienteModel();
		viewPessoaClienteModel?.nome = plutoRow.cells['viewPessoaClienteModel']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		osStatusModel = OsStatusModel();
		osStatusModel?.nome = plutoRow.cells['osStatusModel']?.value;
	}	

	OsAberturaModel clone() {
		return OsAberturaModel(
			id: id,
			idOsStatus: idOsStatus,
			idColaborador: idColaborador,
			idCliente: idCliente,
			numero: numero,
			dataInicio: dataInicio,
			horaInicio: horaInicio,
			dataPrevisao: dataPrevisao,
			horaPrevisao: horaPrevisao,
			dataFim: dataFim,
			horaFim: horaFim,
			nomeContato: nomeContato,
			foneContato: foneContato,
			observacaoCliente: observacaoCliente,
			observacaoAbertura: observacaoAbertura,
			osAberturaEquipamentoModelList: osAberturaEquipamentoModelListClone(osAberturaEquipamentoModelList!),
			osProdutoServicoModelList: osProdutoServicoModelListClone(osProdutoServicoModelList!),
			osEvolucaoModelList: osEvolucaoModelListClone(osEvolucaoModelList!),
		);			
	}

	osAberturaEquipamentoModelListClone(List<OsAberturaEquipamentoModel> osAberturaEquipamentoModelList) { 
		List<OsAberturaEquipamentoModel> resultList = [];
		for (var osAberturaEquipamentoModel in osAberturaEquipamentoModelList) {
			resultList.add(
				OsAberturaEquipamentoModel(
					id: osAberturaEquipamentoModel.id,
					idOsAbertura: osAberturaEquipamentoModel.idOsAbertura,
					idOsEquipamento: osAberturaEquipamentoModel.idOsEquipamento,
					tipoCobertura: osAberturaEquipamentoModel.tipoCobertura,
					numeroSerie: osAberturaEquipamentoModel.numeroSerie,
				)
			);
		}
		return resultList;
	}

	osProdutoServicoModelListClone(List<OsProdutoServicoModel> osProdutoServicoModelList) { 
		List<OsProdutoServicoModel> resultList = [];
		for (var osProdutoServicoModel in osProdutoServicoModelList) {
			resultList.add(
				OsProdutoServicoModel(
					id: osProdutoServicoModel.id,
					idOsAbertura: osProdutoServicoModel.idOsAbertura,
					idProduto: osProdutoServicoModel.idProduto,
					tipo: osProdutoServicoModel.tipo,
					complemento: osProdutoServicoModel.complemento,
					quantidade: osProdutoServicoModel.quantidade,
					valorUnitario: osProdutoServicoModel.valorUnitario,
					valorSubtotal: osProdutoServicoModel.valorSubtotal,
					taxaDesconto: osProdutoServicoModel.taxaDesconto,
					valorDesconto: osProdutoServicoModel.valorDesconto,
					valorTotal: osProdutoServicoModel.valorTotal,
				)
			);
		}
		return resultList;
	}

	osEvolucaoModelListClone(List<OsEvolucaoModel> osEvolucaoModelList) { 
		List<OsEvolucaoModel> resultList = [];
		for (var osEvolucaoModel in osEvolucaoModelList) {
			resultList.add(
				OsEvolucaoModel(
					id: osEvolucaoModel.id,
					idOsAbertura: osEvolucaoModel.idOsAbertura,
					dataRegistro: osEvolucaoModel.dataRegistro,
					horaRegistro: osEvolucaoModel.horaRegistro,
					enviarEmail: osEvolucaoModel.enviarEmail,
					observacao: osEvolucaoModel.observacao,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/ordem_servico/src/model/os_abertura_model.py
from src import db
from src.model.os_abertura_equipamento_model import OsAberturaEquipamentoModel
from src.model.os_produto_servico_model import OsProdutoServicoModel
from src.model.os_evolucao_model import OsEvolucaoModel
from src.model.view_pessoa_cliente_model import ViewPessoaClienteModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.os_status_model import OsStatusModel


class OsAberturaModel(db.Model):
    __tablename__ = 'os_abertura'

    id = db.Column(db.Integer, primary_key=True)
    numero = db.Column(db.String(20))
    data_inicio = db.Column(db.DateTime)
    hora_inicio = db.Column(db.String(8))
    data_previsao = db.Column(db.DateTime)
    hora_previsao = db.Column(db.String(8))
    data_fim = db.Column(db.DateTime)
    hora_fim = db.Column(db.String(8))
    nome_contato = db.Column(db.String(50))
    fone_contato = db.Column(db.String(15))
    observacao_cliente = db.Column(db.Text)
    observacao_abertura = db.Column(db.Text)
    id_cliente = db.Column(db.Integer, db.ForeignKey('view_pessoa_cliente.id'))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_os_status = db.Column(db.Integer, db.ForeignKey('os_status.id'))

    os_abertura_equipamento_model_list = db.relationship('OsAberturaEquipamentoModel', lazy='dynamic')
    os_produto_servico_model_list = db.relationship('OsProdutoServicoModel', lazy='dynamic')
    os_evolucao_model_list = db.relationship('OsEvolucaoModel', lazy='dynamic')
    view_pessoa_cliente_model = db.relationship('ViewPessoaClienteModel', foreign_keys=[id_cliente])
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    os_status_model = db.relationship('OsStatusModel', foreign_keys=[id_os_status])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_os_status = data.get('idOsStatus')
        self.id_colaborador = data.get('idColaborador')
        self.id_cliente = data.get('idCliente')
        self.numero = data.get('numero')
        self.data_inicio = data.get('dataInicio')
        self.hora_inicio = data.get('horaInicio')
        self.data_previsao = data.get('dataPrevisao')
        self.hora_previsao = data.get('horaPrevisao')
        self.data_fim = data.get('dataFim')
        self.hora_fim = data.get('horaFim')
        self.nome_contato = data.get('nomeContato')
        self.fone_contato = data.get('foneContato')
        self.observacao_cliente = data.get('observacaoCliente')
        self.observacao_abertura = data.get('observacaoAbertura')

    def serialize(self):
        return {
            'id': self.id,
            'idOsStatus': self.id_os_status,
            'idColaborador': self.id_colaborador,
            'idCliente': self.id_cliente,
            'numero': self.numero,
            'dataInicio': self.data_inicio.isoformat(),
            'horaInicio': self.hora_inicio,
            'dataPrevisao': self.data_previsao.isoformat(),
            'horaPrevisao': self.hora_previsao,
            'dataFim': self.data_fim.isoformat(),
            'horaFim': self.hora_fim,
            'nomeContato': self.nome_contato,
            'foneContato': self.fone_contato,
            'observacaoCliente': self.observacao_cliente,
            'observacaoAbertura': self.observacao_abertura,
            'osAberturaEquipamentoModelList': [os_abertura_equipamento_model.serialize() for os_abertura_equipamento_model in self.os_abertura_equipamento_model_list],
            'osProdutoServicoModelList': [os_produto_servico_model.serialize() for os_produto_servico_model in self.os_produto_servico_model_list],
            'osEvolucaoModelList': [os_evolucao_model.serialize() for os_evolucao_model in self.os_evolucao_model_list],
            'viewPessoaClienteModel': self.view_pessoa_cliente_model.serialize() if self.view_pessoa_cliente_model else None,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'osStatusModel': self.os_status_model.serialize() if self.os_status_model else None,
        }